otfvis: cannot press "step" while simulation is playing

In this state:
![image](https://user-images.githubusercontent.com/8777649/114306768-35714e00-9add-11eb-8edf-727db7d4990c.png)
(i.e. simulation is running; one could press the pause button) pressing the step button (single triangle) aborts the simulation:
```
...2021-04-11T15:45:41,421  INFO QSim:542 SIMULATION (NEW QSim) AT 04:00:00 : #Veh=288124 lost=0 simT=14400.0s realT=74s; (s/r): 194.59459459459458
2021-04-11T15:45:49,147  INFO AbstractQNetsimEngine:365 SIMULATION (QNetsimEngine) AT 05:00:00 : #links=4244 #nodes=557
2021-04-11T15:45:49,148  INFO QSim:542 SIMULATION (NEW QSim) AT 05:00:00 : #Veh=285190 lost=0 simT=18000.0s realT=82s; (s/r): 219.5121951219512
2021-04-11T15:46:20,287  INFO MemoryObserver:41 used RAM: 3350 MB  free: 649 MB  total: 4000 MB
2021-04-11T15:46:30,548 ERROR MatsimRuntimeModifications$1:62 Getting uncaught Exception in Thread AWT-EventQueue-0
java.lang.IllegalStateException: null
	at org.matsim.core.mobsim.framework.PlayPauseSimulationControl.doStep(PlayPauseSimulationControl.java:87) ~[matsim-13.0-20210407.195228-252.jar:?]
	at org.matsim.vis.otfvis.OnTheFlyServer.doStep(OnTheFlyServer.java:311) ~[otfvis-13.0-20210407.204757-353.jar:?]
	at org.matsim.vis.otfvis.OnTheFlyServer.requestNewTime(OnTheFlyServer.java:321) ~[otfvis-13.0-20210407.204757-353.jar:?]
	at org.matsim.vis.otfvis.gui.OTFHostControl.requestTimeStep(OTFHostControl.java:88) ~[otfvis-13.0-20210407.204757-353.jar:?]
	at org.matsim.vis.otfvis.gui.OTFControlBar$5.actionPerformed(OTFControlBar.java:148) ~[otfvis-13.0-20210407.204757-353.jar:?]
	at javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1967) ~[?:?]
	at javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2308) ~[?:?]
	at javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:405) ~[?:?]
	at javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:262) ~[?:?]
	at javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:279) ~[?:?]
	at java.awt.AWTEventMulticaster.mouseReleased(AWTEventMulticaster.java:298) ~[?:?]
	at java.awt.Component.processMouseEvent(Component.java:6636) ~[?:?]
	at javax.swing.JComponent.processMouseEvent(JComponent.java:3342) ~[?:?]
	at java.awt.Component.processEvent(Component.java:6401) ~[?:?]
	at java.awt.Container.processEvent(Container.java:2263) ~[?:?]
	at java.awt.Component.dispatchEventImpl(Component.java:5012) ~[?:?]
	at java.awt.Container.dispatchEventImpl(Container.java:2321) ~[?:?]
	at java.awt.Component.dispatchEvent(Component.java:4844) ~[?:?]
	at java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4919) ~[?:?]
	at java.awt.LightweightDispatcher.processMouseEvent(Container.java:4548) ~[?:?]
	at java.awt.LightweightDispatcher.dispatchEvent(Container.java:4489) ~[?:?]
	at java.awt.Container.dispatchEventImpl(Container.java:2307) ~[?:?]
	at java.awt.Window.dispatchEventImpl(Window.java:2764) ~[?:?]
	at java.awt.Component.dispatchEvent(Component.java:4844) ~[?:?]
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:772) ~[?:?]
	at java.awt.EventQueue$4.run(EventQueue.java:721) ~[?:?]
	at java.awt.EventQueue$4.run(EventQueue.java:715) ~[?:?]
	at java.security.AccessController.doPrivileged(AccessController.java:391) ~[?:?]
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85) ~[?:?]
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95) ~[?:?]
	at java.awt.EventQueue$5.run(EventQueue.java:745) ~[?:?]
	at java.awt.EventQueue$5.run(EventQueue.java:743) ~[?:?]
	at java.security.AccessController.doPrivileged(AccessController.java:391) ~[?:?]
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85) ~[?:?]
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:742) ~[?:?]
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203) [?:?]
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124) [?:?]
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113) [?:?]
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109) [?:?]
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101) [?:?]
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:90) [?:?]
2021-04-11T15:46:31,244 ERROR AbstractController:222 Mobsim did not complete normally! afterMobsimListeners will be called anyway.
java.lang.RuntimeException: java.lang.InterruptedException
	at org.matsim.core.mobsim.qsim.qnetsimengine.QNetsimEngineWithThreadpool.run(QNetsimEngineWithThreadpool.java:110) ~[matsim-13.0-20210407.195228-252.jar:?]
	at org.matsim.core.mobsim.qsim.qnetsimengine.AbstractQNetsimEngine.doSimStep(AbstractQNetsimEngine.java:225) ~[matsim-13.0-20210407.195228-252.jar:?]
	at org.matsim.core.mobsim.qsim.QSim.doSimStep(QSim.java:395) ~[matsim-13.0-20210407.195228-252.jar:?]
	at org.matsim.core.mobsim.qsim.QSim.run(QSim.java:254) ~[matsim-13.0-20210407.195228-252.jar:?]
	at org.matsim.core.controler.NewControler.runMobSim(NewControler.java:124) ~[matsim-13.0-20210407.195228-252.jar:?]
	at org.matsim.core.controler.AbstractController$8.run(AbstractController.java:211) ~[matsim-13.0-20210407.195228-252.jar:?]
	at org.matsim.core.controler.AbstractController.iterationStep(AbstractController.java:243) ~[matsim-13.0-20210407.195228-252.jar:?]
	at org.matsim.core.controler.AbstractController.mobsim(AbstractController.java:207) [matsim-13.0-20210407.195228-252.jar:?]
	at org.matsim.core.controler.AbstractController.iteration(AbstractController.java:154) [matsim-13.0-20210407.195228-252.jar:?]
	at org.matsim.core.controler.AbstractController.doIterations(AbstractController.java:121) [matsim-13.0-20210407.195228-252.jar:?]
	at org.matsim.core.controler.AbstractController$1.run(AbstractController.java:81) [matsim-13.0-20210407.195228-252.jar:?]
	at org.matsim.core.controler.MatsimRuntimeModifications.run(MatsimRuntimeModifications.java:69) [matsim-13.0-20210407.195228-252.jar:?]
	at org.matsim.core.controler.MatsimRuntimeModifications.run(MatsimRuntimeModifications.java:52) [matsim-13.0-20210407.195228-252.jar:?]
	at org.matsim.core.controler.AbstractController.run(AbstractController.java:89) [matsim-13.0-20210407.195228-252.jar:?]
	at org.matsim.core.controler.NewControler.run(NewControler.java:82) [matsim-13.0-20210407.195228-252.jar:?]
	at org.matsim.core.controler.Controler.run(Controler.java:249) [matsim-13.0-20210407.195228-252.jar:?]
	at playground.kairuns.run.KNTransitControler.main(KNTransitControler.java:160) [classes/:?]
Caused by: java.lang.InterruptedException
	at java.util.concurrent.FutureTask.awaitDone(FutureTask.java:418) ~[?:?]
	at java.util.concurrent.FutureTask.get(FutureTask.java:190) ~[?:?]
	at java.util.concurrent.AbstractExecutorService.invokeAll(AbstractExecutorService.java:247) ~[?:?]
	at org.matsim.core.mobsim.qsim.qnetsimengine.QNetsimEngineWithThreadpool.run(QNetsimEngineWithThreadpool.java:106) ~[matsim-13.0-20210407.195228-252.jar:?]
	... 16 more
2021-04-11T15:46:31,249  WARN CharyparNagelActivityScoring:246 The first and the last activity do not have the same type. 
2021-04-11T15:46:31,249  WARN CharyparNagelActivityScoring:247 Will score the first activity from midnight to its end, and the last activity from its start to midnight.
2021-04-11T15:46:31,249  WARN CharyparNagelActivityScoring:248 Because of the nonlinear function, this is not the same as scoring from start to end.
...
```

--- a/contribs/otfvis/src/main/java/org/matsim/vis/otfvis/gui/OTFControlBar.java
+++ b/contribs/otfvis/src/main/java/org/matsim/vis/otfvis/gui/OTFControlBar.java
@@ -20,52 +20,58 @@
 
 package org.matsim.vis.otfvis.gui;
 
+import java.awt.Dimension;
+import java.awt.event.KeyEvent;
+import java.text.MessageFormat;
+
+import javax.swing.ImageIcon;
+import javax.swing.JButton;
+import javax.swing.JCheckBox;
+import javax.swing.JFormattedTextField;
+import javax.swing.JLabel;
+import javax.swing.JTextField;
+import javax.swing.JToolBar;
+import javax.swing.SwingUtilities;
+
 import org.matsim.core.gbl.MatsimResource;
 import org.matsim.core.utils.misc.Time;
 import org.matsim.vis.otfvis.OTFClientControl;
 import org.matsim.vis.otfvis.interfaces.OTFServer;
 import org.matsim.vis.otfvis.opengl.drawer.OTFOGLDrawer;
 
-import javax.swing.*;
-import javax.swing.event.ChangeEvent;
-import javax.swing.event.ChangeListener;
-import java.awt.*;
-import java.awt.event.*;
-import java.text.MessageFormat;
-
 public final class OTFControlBar extends JToolBar {
 
 	private static final long serialVersionUID = 1L;
 
-	private JButton playPauseButton;
+	private final JButton playPauseButton;
 
-	private JFormattedTextField timeField;
+	private final JFormattedTextField timeField;
 
-	private JFormattedTextField scaleField;
+	private final JFormattedTextField scaleField;
 
 	private final OTFHostControl hostControl;
 
-    private boolean synchronizedPlay = true;
+	private boolean synchronizedPlay = true;
 
-	private ImageIcon playIcon = null;
+	private final ImageIcon playIcon;
 
-	private ImageIcon pauseIcon = null;
+	private final ImageIcon pauseIcon;
 
 	private OTFAbortGoto progressBar = null;
 
 	private enum PlayOrPause {PLAY, PAUSE}
 
-    private final OTFOGLDrawer drawer;
+	private final OTFOGLDrawer drawer;
 
 	/*
 	 * The symbol which the play/pause button currently shows.
 	 * (NOT the mode we are in!)
 	 */
-	private PlayOrPause playOrPause = PlayOrPause.PLAY;
+	private PlayOrPause playOrPauseButtonMode = PlayOrPause.PLAY;
 
-	private OTFServer server;
+	private final OTFServer server;
 
-	public OTFControlBar(OTFServer server, final OTFHostControl hostControl, OTFOGLDrawer mainDrawer)  {
+	public OTFControlBar(OTFServer server, final OTFHostControl hostControl, OTFOGLDrawer mainDrawer) {
 		this.server = server;
 		this.hostControl = hostControl;
 		this.drawer = mainDrawer;
@@ -75,176 +81,139 @@ public final class OTFControlBar extends JToolBar {
 		pauseIcon = new ImageIcon(MatsimResource.getAsImage("otfvis/buttonPause.png"), "Pause");
 
 		if (!server.isLive()) {
-			JButton button2 = new JButton();
-			button2.putClientProperty("JButton.buttonType","icon");
-			button2.addActionListener(new ActionListener() {
-				@Override
-				public void actionPerformed(ActionEvent e) {
-					OTFControlBar.this.hostControl.toStart();
-					repaint();
-				}
+			JButton restartButton = new JButton();
+			restartButton.putClientProperty("JButton.buttonType", "icon");
+			restartButton.addActionListener(e -> {
+				OTFControlBar.this.hostControl.toStart();
+				repaint();
 			});
-			button2.setToolTipText("restart the server/simulation");
+			restartButton.setToolTipText("restart the server/simulation");
 			String imgLocation2 = "otfvis/" + "buttonRestart" + ".png";
 			ImageIcon icon2 = new ImageIcon(MatsimResource.getAsImage(imgLocation2), "Restart");
-			button2.setIcon(icon2);
-			add(button2);
-			JButton button1 = new JButton();
-			button1.putClientProperty("JButton.buttonType","icon");
-			button1.addActionListener(new ActionListener() {
-				@Override
-				public void actionPerformed(ActionEvent e) {
-					int bigStep = OTFClientControl.getInstance().getOTFVisConfig().getBigTimeStep();
-					hostControl.requestTimeStep(OTFControlBar.this.hostControl.getSimTime() - bigStep);
-				}
+			restartButton.setIcon(icon2);
+			add(restartButton);
+			JButton multiStepsBackwardButton = new JButton();
+			multiStepsBackwardButton.putClientProperty("JButton.buttonType", "icon");
+			multiStepsBackwardButton.addActionListener(e -> {
+				int bigStep = OTFClientControl.getInstance().getOTFVisConfig().getBigTimeStep();
+				hostControl.requestTimeStep(OTFControlBar.this.hostControl.getSimTime() - bigStep);
 			});
-			button1.setToolTipText("go several timesteps backwards");
+			multiStepsBackwardButton.setToolTipText("go several timesteps backwards");
 			String imgLocation1 = "otfvis/" + "buttonStepBB" + ".png";
 			ImageIcon icon1 = new ImageIcon(MatsimResource.getAsImage(imgLocation1), "<<");
-			button1.setIcon(icon1);
-			add(button1);
-			JButton button = new JButton();
-			button.putClientProperty("JButton.buttonType","icon");
-			button.addActionListener(new ActionListener() {
-				@Override
-				public void actionPerformed(ActionEvent e) {
-					hostControl.requestTimeStep(OTFControlBar.this.hostControl.getSimTime() - 1);
-				}
-			});
-			button.setToolTipText("go one timestep backwards");
+			multiStepsBackwardButton.setIcon(icon1);
+			add(multiStepsBackwardButton);
+			JButton singleStepBackwardButton = new JButton();
+			singleStepBackwardButton.putClientProperty("JButton.buttonType", "icon");
+			singleStepBackwardButton.addActionListener(
+					e -> hostControl.requestTimeStep(OTFControlBar.this.hostControl.getSimTime() - 1));
+			singleStepBackwardButton.setToolTipText("go one timestep backwards");
 			String imgLocation = "otfvis/" + "buttonStepB" + ".png";
 			ImageIcon icon = new ImageIcon(MatsimResource.getAsImage(imgLocation), "<");
-			button.setIcon(icon);
-			add(button);
+			singleStepBackwardButton.setIcon(icon);
+			add(singleStepBackwardButton);
 		}
 
-		JButton button2 = new JButton();
-		button2.putClientProperty("JButton.buttonType","icon");
-		button2.addActionListener(new ActionListener() {
-			@Override
-			public void actionPerformed(ActionEvent e) {
-				if (playOrPause == PlayOrPause.PLAY) {
-					playPauseButton.setIcon(pauseIcon);
-					OTFControlBar.this.hostControl.play(synchronizedPlay);
-					playOrPause = PlayOrPause.PAUSE;
-				} else if (playOrPause == PlayOrPause.PAUSE) {
-					playPauseButton.setIcon(playIcon);
-					OTFControlBar.this.hostControl.pause();
-					playOrPause = PlayOrPause.PLAY;
-				}
-			}
-		});
-		button2.setToolTipText("press to play simulation continuously");
+		JButton playOrPauseButton = new JButton();
+		playOrPauseButton.putClientProperty("JButton.buttonType", "icon");
+		playOrPauseButton.setToolTipText("press to play simulation continuously");
 		String imgLocation2 = "otfvis/" + "buttonPlay" + ".png";
 		ImageIcon icon2 = new ImageIcon(MatsimResource.getAsImage(imgLocation2), "PLAY");
-		button2.setIcon(icon2);
-		playPauseButton = button2;
-		add(playPauseButton);
-		JButton button1 = new JButton();
-		button1.putClientProperty("JButton.buttonType","icon");
-		button1.addActionListener(new ActionListener() {
-			@Override
-			public void actionPerformed(ActionEvent e) {
-				hostControl.requestTimeStep(OTFControlBar.this.hostControl.getSimTime() + 1);
-			}
-		});
-		button1.setToolTipText("go one timestep forward");
+		playOrPauseButton.setIcon(icon2);
+		this.playPauseButton = playOrPauseButton;
+		add(this.playPauseButton);
+		JButton singleStepForwardButton = new JButton();
+		singleStepForwardButton.putClientProperty("JButton.buttonType", "icon");
+		singleStepForwardButton.addActionListener(
+				e -> hostControl.requestTimeStep(OTFControlBar.this.hostControl.getSimTime() + 1));
+		singleStepForwardButton.setToolTipText("go one timestep forward");
 		String imgLocation1 = "otfvis/" + "buttonStepF" + ".png";
 		ImageIcon icon1 = new ImageIcon(MatsimResource.getAsImage(imgLocation1), ">");
-		button1.setIcon(icon1);
-		add(button1);
-		JButton button = new JButton();
-		button.putClientProperty("JButton.buttonType","icon");
-		button.addActionListener(new ActionListener() {
-			@Override
-			public void actionPerformed(ActionEvent e) {
-				int bigStep = OTFClientControl.getInstance().getOTFVisConfig().getBigTimeStep();
-				hostControl.requestTimeStep(OTFControlBar.this.hostControl.getSimTime() + bigStep);
-			}
+		singleStepForwardButton.setIcon(icon1);
+		add(singleStepForwardButton);
+		JButton multiStepForwardButton = new JButton();
+		multiStepForwardButton.putClientProperty("JButton.buttonType", "icon");
+		multiStepForwardButton.addActionListener(e -> {
+			int bigStep = OTFClientControl.getInstance().getOTFVisConfig().getBigTimeStep();
+			hostControl.requestTimeStep(OTFControlBar.this.hostControl.getSimTime() + bigStep);
 		});
-		button.setToolTipText("go several timesteps forward");
+		multiStepForwardButton.setToolTipText("go several timesteps forward");
 		String imgLocation = "otfvis/" + "buttonStepFF" + ".png";
 		ImageIcon icon = new ImageIcon(MatsimResource.getAsImage(imgLocation), ">>");
-		button.setIcon(icon);
-		add(button);
+		multiStepForwardButton.setIcon(icon);
+		add(multiStepForwardButton);
+
+		playOrPauseButton.addActionListener(e -> {
+			if (playOrPauseButtonMode == PlayOrPause.PLAY) {
+				this.playPauseButton.setIcon(pauseIcon);
+				//PlayPauseSimulationControl.doStep() cannot be called while playing in async mode
+				singleStepForwardButton.setEnabled(synchronizedPlay);
+				multiStepForwardButton.setEnabled(synchronizedPlay);
+				OTFControlBar.this.hostControl.play(synchronizedPlay);
+				playOrPauseButtonMode = PlayOrPause.PAUSE;
+			} else if (playOrPauseButtonMode == PlayOrPause.PAUSE) {
+				this.playPauseButton.setIcon(playIcon);
+				OTFControlBar.this.hostControl.pause();
+				playOrPauseButtonMode = PlayOrPause.PLAY;
+				singleStepForwardButton.setEnabled(true);
+				multiStepForwardButton.setEnabled(true);
+			}
+		});
+
 		MessageFormat format = new MessageFormat("{0,number,00}:{1,number,00}:{2,number,00}");
 		timeField = new JFormattedTextField(format);
-		timeField.setMaximumSize(new Dimension(100,30));
-		timeField.setMinimumSize(new Dimension(80,30));
+		timeField.setMaximumSize(new Dimension(100, 30));
+		timeField.setMinimumSize(new Dimension(80, 30));
 		timeField.setHorizontalAlignment(JTextField.CENTER);
-		add( timeField );
-		timeField.addActionListener(new ActionListener() {
-			@Override
-			public void actionPerformed(ActionEvent e) {
-				final int newTime_s = (int) Time.parseTime(timeField.getText());
-				progressBar = new OTFAbortGoto(OTFControlBar.this.server, newTime_s);
-				progressBar.start();
-				new Thread() {
-					@Override
-					public void run() {
-						OTFControlBar.this.hostControl.requestTimeStep(newTime_s);
-						progressBar.terminate = true;
-					}
-				}.start();
-			}
+		add(timeField);
+		timeField.addActionListener(e -> {
+			final int newTime_s = (int)Time.parseTime(timeField.getText());
+			progressBar = new OTFAbortGoto(OTFControlBar.this.server, newTime_s);
+			progressBar.start();
+			new Thread(() -> {
+				OTFControlBar.this.hostControl.requestTimeStep(newTime_s);
+				progressBar.terminate = true;
+			}).start();
 		});
-		hostControl.getSimTimeModel().addChangeListener(new ChangeListener() {
-			@Override
-			public void stateChanged(ChangeEvent e) {
-				SwingUtilities.invokeLater(new Runnable() {
-					@Override
-					public void run() {
-						timeField.setText(Time.writeTime(hostControl.getSimTime()));
-					}
-				});
-			}
-		});
+		hostControl.getSimTimeModel()
+				.addChangeListener(e -> SwingUtilities.invokeLater(
+						() -> timeField.setText(Time.writeTime(hostControl.getSimTime()))));
 		timeField.setText(Time.writeTime(hostControl.getSimTime()));
 
 		if (this.server.isLive()) {
 			final JCheckBox synchBox = new JCheckBox("Synch");
 			synchBox.setMnemonic(KeyEvent.VK_V);
 			synchBox.setSelected(synchronizedPlay);
-			synchBox.addItemListener(new ItemListener() {
-				@Override
-				public void itemStateChanged(ItemEvent e) {
-					OTFControlBar.this.hostControl.pause();
-					synchronizedPlay = synchBox.isSelected();
-					if (playOrPause == PlayOrPause.PAUSE) { // means: playing
-						OTFControlBar.this.hostControl.play(synchronizedPlay);
-					}
+			synchBox.addItemListener(e -> {
+				OTFControlBar.this.hostControl.pause();
+				synchronizedPlay = synchBox.isSelected();
+				if (playOrPauseButtonMode == PlayOrPause.PAUSE) { // means: playing
+					//PlayPauseSimulationControl.doStep() cannot be called while playing in async mode
+					singleStepForwardButton.setEnabled(synchronizedPlay);
+					multiStepForwardButton.setEnabled(synchronizedPlay);
+					OTFControlBar.this.hostControl.play(synchronizedPlay);
 				}
 			});
 			add(synchBox);
 		}
 
 		JLabel lab = new JLabel("Scale: ");
-		lab.setMaximumSize(new Dimension(100,30));
-		lab.setMinimumSize(new Dimension(80,30));
+		lab.setMaximumSize(new Dimension(100, 30));
+		lab.setMinimumSize(new Dimension(80, 30));
 		lab.setHorizontalAlignment(JLabel.RIGHT);
 		add(lab);
 		scaleField = new JFormattedTextField("1.0");
-		scaleField.setMaximumSize(new Dimension(50,30));
-		scaleField.setMinimumSize(new Dimension(30,30));
+		scaleField.setMaximumSize(new Dimension(50, 30));
+		scaleField.setMinimumSize(new Dimension(30, 30));
 		scaleField.setHorizontalAlignment(JTextField.CENTER);
-		add( scaleField );
-		scaleField.addActionListener(new ActionListener() {
-			@Override
-			public void actionPerformed(ActionEvent e) {
-				drawer.setScale(Float.parseFloat(scaleField.getText()));
-			}
-		});
-		drawer.addChangeListener(new ChangeListener() {
-			@Override
-			public void stateChanged(ChangeEvent e) {
-				updateScaleField();
-			}
-		});
+		add(scaleField);
+		scaleField.addActionListener(e -> drawer.setScale(Float.parseFloat(scaleField.getText())));
+		drawer.addChangeListener(e -> updateScaleField());
 		updateScaleField();
 	}
 
 	private void updateScaleField() {
-		scaleField.setText(String.valueOf((double) ((float)(Math.round(drawer.getScale()*100))/100)));
+		scaleField.setText(String.valueOf((double)((float)(Math.round(drawer.getScale() * 100)) / 100)));
 	}
 
 }

--- a/contribs/otfvis/src/main/java/org/matsim/vis/otfvis/gui/OTFHostControl.java
+++ b/contribs/otfvis/src/main/java/org/matsim/vis/otfvis/gui/OTFHostControl.java
@@ -66,12 +66,7 @@ public class OTFHostControl implements GLEventListener {
 			// Live mode without timesteps
 			simTime = new DefaultBoundedRangeModel(0 /* value */, 0 /* extent */, 0 /* value */, Integer.MAX_VALUE /* max */);
 		}
-		simTime.addChangeListener(new ChangeListener() {
-			@Override
-			public void stateChanged(ChangeEvent e) {
-				canvas.repaint();
-			}
-		});
+		simTime.addChangeListener(e -> canvas.repaint());
 		animator = new Animator();
 		animator.add(((GLAutoDrawable) canvas));
 		((GLAutoDrawable) canvas).addGLEventListener(this);
@@ -105,12 +100,7 @@ public class OTFHostControl implements GLEventListener {
 		log.debug("Pressed PLAY, creating animator.");
 		playing = true;
 		syncronizedPlay = synchronizedPlay;
-		new Thread(new Runnable() {
-			@Override
-			public void run() {
-				animator.start();
-			}
-		}).start();
+		new Thread(animator::start).start();
 		if (!synchronizedPlay) {
 			((OTFLiveServer) server).play();
 		}

--- a/matsim/src/main/java/org/matsim/core/mobsim/framework/PlayPauseSimulationControl.java
+++ b/matsim/src/main/java/org/matsim/core/mobsim/framework/PlayPauseSimulationControl.java
@@ -29,9 +29,9 @@ import org.matsim.core.mobsim.framework.listeners.MobsimBeforeCleanupListener;
 import org.matsim.core.mobsim.framework.listeners.MobsimBeforeSimStepListener;
 
 /**
- * Extracted the play/pause functionality from otfvis to make it available for other purposes (specifically, RMITs 
+ * Extracted the play/pause functionality from otfvis to make it available for other purposes (specifically, RMITs
  * CombineSim approach).
- * 
+ *
  * @author nagel
  */
 public class PlayPauseSimulationControl implements PlayPauseSimulationControlI {
@@ -56,7 +56,8 @@ public class PlayPauseSimulationControl implements PlayPauseSimulationControlI {
 		qSim.addQueueSimulationListeners(playPauseMobsimListener);
 	}
 
-	private class PlayPauseMobsimListener implements MobsimBeforeSimStepListener, MobsimAfterSimStepListener, MobsimBeforeCleanupListener {
+	private class PlayPauseMobsimListener
+			implements MobsimBeforeSimStepListener, MobsimAfterSimStepListener, MobsimBeforeCleanupListener {
 		@Override
 		public void notifyMobsimBeforeSimStep(MobsimBeforeSimStepEvent event) {
 			try {
@@ -65,15 +66,17 @@ public class PlayPauseSimulationControl implements PlayPauseSimulationControlI {
 				throw new RuntimeException(e);
 			}
 		}
+
 		@Override
 		public void notifyMobsimAfterSimStep(MobsimAfterSimStepEvent event) {
 			access.release();
-			localTime = (int) event.getSimulationTime();
+			localTime = (int)event.getSimulationTime();
 			// yy I am not so sure about the "int".  kai, nov'17
 			stepDone.arriveAndAwaitAdvance();
 			// This is arrival by one party.  If "pause" has been pressed before, we have a second party, and thus do not
 			// advance.
 		}
+
 		@Override
 		public void notifyMobsimBeforeCleanup(MobsimBeforeCleanupEvent e) {
 			localTime = Double.MAX_VALUE;
@@ -84,6 +87,8 @@ public class PlayPauseSimulationControl implements PlayPauseSimulationControlI {
 	@Override
 	public final void doStep(int time) {
 		if (status == Status.PLAY) {
+			// this may happen when clicking single/multi-step forward buttons while playing in the async mode,
+			// so that combination should be disabled in GUI, michalm
 			throw new IllegalStateException();
 		}
 		while (localTime < time) {
@@ -117,7 +122,7 @@ public class PlayPauseSimulationControl implements PlayPauseSimulationControlI {
 	}
 
 	public final boolean isFinished() {
-		return ( localTime == Double.MAX_VALUE ) ;
+		return (localTime == Double.MAX_VALUE);
 	}
 
 	public final double getLocalTime() {

